ci: publish docs automatically

nealvaidya opened the pull request:
#### Overview:

This PR modifies the generate-docs workflow in GitHub Actions to publish docs automatically to s3 when docs are successfully built under certain conditions. 
For any push to the `main` branch, it will publish to the `dev` docs path. (this is commented out so we can test incrementally) 
For tags, it will publish to a subpath with the same name as the tag, as well as to the `latest` docs path. (the `latest` part is commented out so we can test incrementally) 
Other commits will not cause publication.


The workflow will also flush the CDN cache upon publication of docs.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

**Chores**
- Enhanced documentation publishing workflow with automated cloud storage deployment, integrated content delivery network cache management, and version handling capabilities
- Added workflow templates and infrastructure improvements to streamline documentation publishing and multi-channel distribution processes
- Improved version management infrastructure with manifest updates and better release tracking across distribution channels

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nealvaidya committed:
ci: add docs publication step

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
ci: build docs on tags

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
add dev (main) and latest (stable) docs paths

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
add copywrite header

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

mikemckiernan reviewed (commented):
Thanks for putting this together.  One other thought (not mine, wish it was) is to find a way for a _dry run_ that describes what action would be taken.  I'll understand if you don't have time to poke at that.

nealvaidya committed:
test: add branch trigger for testing publish-s3 workflow

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
modify variable name

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
make s3 bucket secret

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
test: add test branch to generate-docs workflow

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
simplify artifact handling

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
fix: handle workflow_dispatch trigger for artifact download

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
test: allow manual publish to dev from test branch

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
hide bucket name

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
feat: merge publish-s3 into generate-docs workflow

- Combine build and publish into single workflow with two jobs
- Simplify artifact handling by using job dependency instead of workflow_run
- Eliminate ~150 lines of artifact searching and waiting logic
- Update workflow name to 'Generate and Publish Documentation'
- Add comprehensive header documentation
- Add workflow_dispatch trigger with version input
- Publish job skips for PRs, runs for main/tags/manual triggers
- Keep test branch support (nealv/docs_build) for now

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
remove standalone publish workflow

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
comments for publish workflow

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
update publish to latest logic

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
update publish to latest logic

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
update publish to latest logic

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
comment out publish to latest and dev for testing

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
comment out latest/dev outputs

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
remove the special case for my branch

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya committed:
prefix akamai vars

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nv-tusharma commented on .github/workflows/generate-docs.yml:
```diff
@@ -159,3 +187,352 @@ jobs:
+
+          echo "published=true" >> "$GITHUB_OUTPUT"
+
+      # - name: Publish latest
```
Is there a reason why we're keeping this commented out?

nealvaidya commented:
> why do we need to publish to s3? Github allows generating a docs website. The link should be posted as a comment to easily access the documentation.
> 
> example - [#4426 (comment)](https://github.com/ai-dynamo/dynamo/pull/4426#issuecomment-3545139254)

This is so we can get the documentation pushed to docs.nvidia.com when we create a new release, and so that the “dev” path in docs.nvidia.com always reflects the latest top of tree docs. I’m not making any changes to the per PR documentation as in your link

nvda-mesharma commented:
> > why do we need to publish to s3? Github allows generating a docs website. The link should be posted as a comment to easily access the documentation.
> > example - [#4426 (comment)](https://github.com/ai-dynamo/dynamo/pull/4426#issuecomment-3545139254)
> 
> This is so we can get the documentation pushed to docs.nvidia.com when we create a new release, and so that the “dev” path in docs.nvidia.com always reflects the latest top of tree docs. I’m not making any changes to the per PR documentation as in your link

thanks for explaining. Is that the standard process to upload to s3 and then push to docs.nvidia.com?

nv-tusharma commented on .github/workflows/generate-docs.yml:
```diff
@@ -159,3 +187,352 @@ jobs:
+          role-to-assume: ${{ secrets.DOCS_AWS_IAM_STS_ROLE }}
+          aws-access-key-id: ${{ secrets.DOCS_AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.DOCS_AWS_SECRET_ACCESS_KEY }}
+          aws-region: ${{ secrets.DOCS_AWS_REGION || 'us-east-1' }}
```
nit, but probably don't want to default aws region if your role isn't setup there.

nv-tusharma commented on .github/workflows/generate-docs.yml:
```diff
@@ -159,3 +187,352 @@ jobs:
+            fi
+
+            # Check for /not-latest flag in commit message
+            if [[ "${COMMIT_MSG}" =~ /not-latest ]]; then
```
Just so I understand this correctly, is this looking for a /not-latest flag in COMMIT_MSG? Does this get auto-generated?

nealvaidya replied:
Yes, that's what it's looking for. It's not autogenerated, you'd need to include it yourself which admittedly is a bit cumbersome. But this is an escape hatch, not the normal behavior

nealvaidya commented:
@nvda-mesharma Yes, docs.nvidia.com should be automatically populated from the S3 bucket we specified in the secrets. Not sure about standard but it’s a process we’ve used internally via gitlab, never GitHub. Hence the incremental initial PR

nealvaidya committed:
remove default region

Signed-off-by: Neal Vaidya <nealv@nvidia.com>

nealvaidya commented:
/ok to test 5a2892e9626f8fb1d98c0906f03c2aac98b74e59
